org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/JDTLanguageServer.java

```diff
@@ -319,9 +330,83 @@ public void didChangeConfiguration(DidChangeConfigurationParams params) {
 			JavaLanguageServerPlugin.logException(e.getMessage(), e);
 		}
 		pm.registerWatchers();
+		configureFormatter();
 		logInfo(">>New configuration: " + settings);
 	}
 
+	public void configureFormatter() {
+		String formaterUrl = preferenceManager.getPreferences().getFormatterUrl();
+		Map<String, String> options = null;
+		if (formaterUrl != null) {
+			URL url = null;
+			try {
+				url = new URL(formaterUrl);
+			} catch (MalformedURLException e1) {
+				File file = new File(formaterUrl);
+				if (file.isFile()) {
+					try {
+						url = file.toURI().toURL();
+					} catch (MalformedURLException e) {
+						JavaLanguageServerPlugin.logInfo("Invalid formatter:" + formaterUrl);
+					}
+				} else {
+					unregisterWatcher(file);
+				}
+			}
+			if (url != null) {
+				try (InputStream inputStream = url.openStream()) {
+					InputSource inputSource = new InputSource(inputStream);
+					String profileName = preferenceManager.getPreferences().getFormatterProfileName();
+					options = ProfileStore.readSettingsFromStream(inputSource, profileName);
+				} catch (Exception e) {
+					JavaLanguageServerPlugin.logException(e.getMessage(), e);
+				} finally {
+					File file = null;
+					try {
+						file = new File(url.toURI());
+					} catch (URISyntaxException e) {
+						file = new File(url.getPath());
+					}
+					if (file != null) {
+						DidChangeWatchedFilesRegistrationOptions didChangeWatchedFilesRegistrationOptions = unregisterWatcher(file);
+						if (file.isFile()) {
+							logInfo(">> registerFeature 'workspace/didChangeWatchedFiles'");
+							JavaLanguageServerPlugin.getInstance().registerCapability(Preferences.WORKSPACE_WATCHED_FILES_ID, Preferences.WORKSPACE_WATCHED_FILES, didChangeWatchedFilesRegistrationOptions);
+						}
+					}
+				}
+			} else {
+				JavaLanguageServerPlugin.logInfo("Invalid formatter:" + formaterUrl);
+			}
+		}
+		if (options != null && !options.isEmpty()) {
+			setFormattingOptions(options);
+		} else {
+			options = DefaultCodeFormatterOptions.getEclipseDefaultSettings().getMap();
+			setFormattingOptions(options);
+			JavaLanguageServerPlugin.getPreferencesManager().initialize();
+		}
+	}
+
+	private void setFormattingOptions(Map<String, String> options) {
+		Hashtable<String, String> javaOptions = JavaCore.getOptions();
+		Set<String> keys = options.keySet();
+		for (String key : keys) {
+			javaOptions.put(key, options.get(key));
```

Comment:
when iterating over all key/value pairs of a map, perform a lookup in each iteration is inefficient.
Please iterate over the entrySet instead or, using a lambda:

```java
		options.forEach((k, v)->{
			javaOptions.put(k, v);
		});
```